### ISSUE TITLE
Readme improvements

### ISSUE THREAD

**Issue by greenfork** [ag91/moldable-emacs#27]
Hi, I have read several blog posts about this package and watched several videos and I still have hard time understanding what this package does. There were several good pieces of advice on [reddit](https://www.reddit.com/r/emacs/comments/qzwxd6/moldable_emacs_lets_make_english_easy_to_query/hlr10oi/) which are mostly related to blog posts. I'd like to help making this readme more understandable by providing some observations I have.

1. The word "story" is over-used. It is used in [BDD](https://en.wikipedia.org/wiki/Behavior-driven_development), [Storybook](https://storybook.js.org/) and probably in more places. The common thing I find across all of them is that the word "story" is a bad analogy for technical people which makes it hard to understand what it really does, but apparently it is a very easy to grasp analogy for non-technical people who don't need to understand what it really is. Current readme seems to use this word in a similar fashion "You have a story in mind", with an example story “Given a code buffer, I want to see duplicated code because I copy-pasted a lot and I want to cleanup.”. If we were to give a definition of this sample "story", I would say it is "What I want to do" + "What I need to have already" or in technical terms function + precondition. Immediate solution might be to remove the word "story" and describe it without any analogies, just say what it really is.
2. I think an important question the readme should address is how this package is better than other similar solutions. What even are similar solutions? Readme should specify similar concepts so we, humans, can build our understanding upon some previous knowledge we have with similar tools. I don't believe this package realizes a unique idea that has never been invented before even approximately or badly.
3. I can summarize my current misunderstanding in a simple question: "how is this different from just a function with a precondition?". I will provide some examples from the readme:
   - "Molds “mold” the data in your context into the data of your result" - this is exactly what a pure function does, transforms data from one form to another form
   - "The power of moldable-emacs is to make this micro-tools highly composable" - pure functions are highly composable, probably no need to comment further

I used the term "pure function" because it is not strictly true for non-pure functions. But if we speak in terms of Emacs, "creating a buffer with some text" might as well qualify as "pure" in certain contexts, and I believe this is also the main point of this package that we use Emacs buffers as input and output of the functions.

**Comment by ag91** [comment]
Thanks for your observations: I am happy to improve things.
1. I use story because is where I want to go towards with this mode: I want to easily communicate meaning supported by moldable-emacs. What moldable-emacs does is providing you with views of some context (by default a buffer, but often even just a selection of text). I call these views molds and they can be combined to produce more complex views to communicate more complex stories to others. If one day Emacs could write the molds for me and I could focus only on the story I want to say, I would be happy.
2. I need help with this. My intuition is that other modes focus on a single view of something (e.g., Magit presents you git data in a way you can easily manipulate), while moldable-emacs tries to make inexpensive to create custom views and compose them. 
3. Not different at all. That is roughly the way it is implemented. What I want to focus though is what you can express with its features. This moldable-emacs is a framework to create molds (pure functions) so that you can easily compose them. It comes with examples that I use often (I use the mold CodeAsTree to tell a story to myself like "Given this code, I have to refactor it because I got a lot of duplication" by viewing a source file as a todo list of the duplicated code elements).

I am pretty sure I am confusing people at the moment and you could view this mode differently. For example, in https://github.com/ag91/moldable-emacs/issues/11 you can find

> That's certainly a great use. I suggest that it can also be a tool to allow user's to solve specific problems they might have by solving general problems in molds. E.g. your excellent blog post about moving private functions to the bottom. I hate having to solve the same problem over and over again in different contexts. I would much rather solve the general problem once, and then be able to use it in many contexts.

I totally agree with her. You can use moldable-emacs to solve problems in a generic way. I see this as included: if I am pair-programming with you and I am telling you this story "Given these 100 files that have a misspelt variable in one of their else statement, we would take forever to do it by hand. But we can do it in in one swoop if we define the correct pattern!", I can actually solve a boring problem in a generic way through views/molds. I can also give to you the "story" by serializing the views I used and their contents (I didn't work on this, but is simple to achieve).

In summary: moldable-emacs aims to bring your focus on what you want to communicate to others (yourself included) by making that simple to express (with molds). 

-- Probably this message will not make things better, sigh..

**Comment by ag91** [comment]
Thanks for that. Sorry for the late reply, your comment is full of interesting points and I am still thinking about all of that :)

I understand where you are coming from. Definitely the story bit is a stretch: the tension comes from what I want this mode to be and the current status of things.

I like your description, I will think some more and then I can try to inject it in the README.

About "bring your focus on what you want to communicate": when I wrote that I was focusing on the presentation aspect not the extension aspect (i.e, how to write a new mold). 90% of the time I work with moldable-emacs I focus on taking some data and making it easy to read for me (or others). I do that mostly by composing existing molds. Only a 10% of time I write a new mold for myself.

For example, I have a mold to list live production errors and I show them to myself grouped by categories and (when I can) with links to the failing source code. This is a view target to me, a developer. 
Still these categories of errors are kept in an Elisp plist. The strength of the presentation layer of moldable-emacs is that I can now transform that plist (targeting a developer) into a bar chart with a pre-existing mold (no code to write, it's immediate), which could make it easy to communicate to management for example. So I can communicate to different audiences the same data without needed new code or integration (unless I need a completely new way of seeing things).

**Comment by greenfork** [comment]
Thanks! I don't think I have something to add to the discussion, so feel free to close this issue. I will also think about this unique concept, that is an interesting approach to information and representation!

**Comment by ag91** [comment]
Thanks again for your help. I have used your snippet and I have rewritten the README once more. Surely requires some more work but I hope it got closer to be understandable :)
